src/control/zoom/ZoomControl.cpp

```diff
@@ -26,6 +26,33 @@ auto onScrolledwindowMainScrollEvent(GtkWidget* widget, GdkEventScroll* event, Z
     return zoom->isZoomPresentationMode();
 }
 
+auto onTouchpadPinchEvent(GtkWidget* widget, GdkEventTouchpadPinch* event, ZoomControl* zoom) -> bool {
+    if (event->type == GDK_TOUCHPAD_PINCH && event->n_fingers == 2) {
+        utl::Point<double> center;
+        switch (event->phase) {
+            case GDK_TOUCHPAD_GESTURE_PHASE_BEGIN:
+                if (zoom->isZoomFitMode()) {
+                    zoom->setZoomFitMode(false);
+                }
+                center = {event->x, event->y};
+                // Need to adjust the center because the event coordinates are relative
+                // to the widget, not to the zoomed (and translated) view
+                center += {zoom->getVisibleRect().x, zoom->getVisibleRect().y};
+                zoom->startZoomSequence(center);
+                break;
+            case GDK_TOUCHPAD_GESTURE_PHASE_UPDATE:
+                zoom->zoomSequenceChange(event->scale, true);
+                break;
+            case GDK_TOUCHPAD_GESTURE_PHASE_END:
+            case GDK_TOUCHPAD_GESTURE_PHASE_CANCEL:
```

Comment:
If the gesture is canceled, the zoom should actually be reverted and not finished. Canceling might be triggered, when GTK decides this is actually a different event.

Reply:
So in the cancel case I should just call zoomSequenceUpdate with a scale of 1 and end the sequence after that, right?
I'm not sure whether this is better from a users perspective - I wouldn't want anything jumping around, that's why I just ended the sequence for the cancel event, too. Not sure how late in the sequence it's possible that the event is cancelled though.

Reply:
Please add a method `cancelZoomSequence` in [ZoomControl.cpp](https://github.com/xournalpp/xournalpp/blob/master/src/control/zoom/ZoomControl.cpp) which reverts the zoom sequence. If I'm not mistaken all the necessary data to revert it fully is already available in this class.

Reply:
Alright, I hope that's fine now. Thanks for your comments!